src/System.Private.CoreLib/shared/System/SpanHelpers.Byte.cs

```diff
@@ -990,24 +1000,122 @@ public static unsafe bool SequenceEqual(ref byte first, ref byte second, nuint l
                     {
                         goto NotEqual;
                     }
+
                     i += sizeof(UIntPtr);
                 }
-                return Unsafe.ReadUnaligned<UIntPtr>(ref Unsafe.AddByteOffset(ref first, n)) ==
-                       Unsafe.ReadUnaligned<UIntPtr>(ref Unsafe.AddByteOffset(ref second, n));
+
+                if (Unsafe.ReadUnaligned<UIntPtr>(ref Unsafe.AddByteOffset(ref first, n)) ==
+                       Unsafe.ReadUnaligned<UIntPtr>(ref Unsafe.AddByteOffset(ref second, n)))
+                {
+                    goto Equal;
+                }
+
+                goto NotEqual;
             }
 
             while ((byte*)n > (byte*)i)
             {
                 if (Unsafe.AddByteOffset(ref first, i) != Unsafe.AddByteOffset(ref second, i))
+                {
                     goto NotEqual;
+                }
+
                 i += 1;
             }
 
-        Equal:
-            return true;
+            goto Equal;
 
-        NotEqual: // Workaround for https://github.com/dotnet/coreclr/issues/13549
-            return false;
+        Avx2:
+            if (Avx2.IsSupported)
```

Comment:
Would putting this in a local function be better?

Reply:
Doubt it, would be changing a `jb       SHORT ` to a `call`

Reply:
Or are you thinking the dropping the two `vzeroupper` calls in the short paths would make up for it?

Reply:
I was just thinking about this from the perspective of making the code more readable (and possibly throughput, since the method is getting overly large and the JIT will have to reason about all the dead code, etc).

Reply:
The dead code part shouldn't be an issue for the JIT, but the `vzeroupper` concern is one motivation, and it *might* generate better code, as the JIT often does worse on large methods (even if they're not so large as to trigger minopts)

Reply:
> Hmm vzeroupper is kinda expensive

`vzeroupper` is a zero-latency instruction, so the only concern is code size overhead.

Reply:
Does `vzeroupper` take 4 cycles though?
![image](https://user-images.githubusercontent.com/1142958/51288159-d2260d80-19f2-11e9-9009-b0d17d159765.png)
So best to move it out of line in the non-vector portion (where length < vec.Length)

Reply:
> Does vzeroupper take 4 cycles though?

No, `vzeroupper` is only expensive on Knights Landing (Xeon Phi). It is 0-latency and 1-throughput on Core/Xeon (Sandybridge, Haswell, Skylake, etc.).